### DISCUSSION TITLE
Empty trash folder

### DISCUSSION THREAD

**Discussion by lonix1** [jstedfast/MailKit#1866] [Q&A] [answered]
I read the [README](https://github.com/jstedfast/MailKit/blob/80dd3176ae23e36208e2e2b54371021dc0ba97bb/README.md#deleting-messages-in-imap) and various other discussions. I understand that IMAP "deletion" is a two-step process: 1) mark message as deleted, and 2) expunge.

On my (non-Gmail) IMAP server, messages are moved to a "Trash" folder rather than being deleted. If I want to then truly delete them, I can "empty" that folder using the server's web interface or a local mail client like Outlook/Thunderbird.

(This IMAP server has a setting to immediately delete rather than use the trash folder. I prefer not to use that though, unless there's no choice.)

Is it possible to perform the "emptying" action using MailKit? (If I empty trash using Thunderbird, that folder is emptied on the server too, so it should be possible.)

**Comment by lonix1** [comment]
UPDATE: After some trial-and-error I found a way to "empty" the trash.

- As shown in the README, open Inbox folder, delete messages and expunge folder
- Open the "Trash" special folder (if it exists), then find those messages again (because I'm unsure whether I can use the same uids from before), mark them as deleted, and expunge that folder

Is this the normal approach? (I assume it won't work on all IMAP servers.)

**Reply by jstedfast** [reply, marked as answer]
>  (because I'm unsure whether I can use the same uids from before)

They will not have the same UIDs.

> Is this the normal approach? (I assume it won't work on all IMAP servers.)

If it works, then it's the right approach :-)

This is a non-standard IMAP server behavior so it's a bit "whatever works"

**Reply by lonix1** [reply]
> is a non-standard IMAP server behavior

Just to be certain, do you mean that "standard" behaviour is a delete+expunge performs a permanent delete (no trash folder involved)?

**Reply by jstedfast** [reply]
Correct. The IMAP protocol does not define any Trash folder at all.

**Reply by lonix1** [reply]
The number of workarounds, gotchas and kludges in SMTP/POP/IMAP is enough to make one's head spin. And some of the RFCs are decades old. I don't understand how you can hold all that arcane knowledge in your head!

This library is incredible, thanks for maintaining it. And all the best for 2025!

**Reply by jstedfast** [reply]
Been working on email-related stuff for over 25 years, so it has been mostly internalized at this point. Probably started back in '98 or '99 on implementing my first email client.
